### PR TITLE
Initialize SSESession if already in connected mode when loaded

### LINKED ISSUE #3668: Initialize SSESession if already in connected mode when package loads
### Description

Due to the async nature of the packages load order, it's possible that when this module starts, we're already in connected mode and missed the event. So we need to also explicitly check in the ctor if we're in connected mode, and if so, start a new session.

### PR DESCRIPTION
**Pull request by georgii-borovinskikh-sonarsource**
Closes #3668

### CODE REVIEW

File: src/TestInfrastructure/Helpers/MefTestHelpers.cs [hunk 1]
```diff
@@ -33,9 +33,9 @@ namespace SonarLint.VisualStudio.Integration.UnitTests
     public static class MefTestHelpers
     {
         [DebuggerStepThrough]
-        public static Export CreateExport<T>() where T: class
+        public static Export CreateExport<T>(T value = null) where T : class
```
Comment (rita-sonarsource) [inline src/TestInfrastructure/Helpers/MefTestHelpers.cs:36]: Why was this needed? there is a method below that accepts a value

File: src/ConnectedMode/ServerSentEvents/SSESessionManager.cs [hunk 2]
```diff
@@ -36,10 +36,12 @@ internal interface ISSESessionManager : IDisposable
     [PartCreationPolicy(CreationPolicy.Shared)]
     internal sealed class SSESessionManager : ISSESessionManager
     {
+        private readonly object syncRoot = new object();
```
Comment (rita-sonarsource) [inline src/ConnectedMode/ServerSentEvents/SSESessionManager.cs:39]: It's a singleton but still better to be explicit and mark it static.

File: src/TestInfrastructure/Helpers/MefTestHelpers.cs (hunk 1 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/TestInfrastructure/Helpers/MefTestHelpers.cs:36]: Hmm, true, I just got thrown off by the fact that that method received an object. Will revert this change

File: src/ConnectedMode/ServerSentEvents/SSESessionManager.cs [hunk 3]
```diff
@@ -67,25 +71,49 @@ public void Dispose()
...
+                currentSession = sessionToLaunch = sseSessionFactory.Create(bindingConfiguration.Project.ProjectKey);
+            }
+
+            try
+            {
+                sessionToLaunch.PumpAllAsync().Forget();
+            }
+            catch (ObjectDisposedException)
+            {
+                // ignore
```
Comment (georgii-borovinskikh-sonarsource) [inline src/ConnectedMode/ServerSentEvents/SSESessionManager.cs:106]: @rita-sonarsource I've only just realized that because the manager doesn't await the Task, it doesn't need the catch

File: src/ConnectedMode/ServerSentEvents/SSESessionManager.cs (hunk 2 above)
Comment (rita-sonarsource) [reply src/ConnectedMode/ServerSentEvents/SSESessionManager.cs:39]: The fields can't be static because they're MEF imported.
I think that actually avoiding deadlocks is a problem -- I'd rather get the deadlock and see it so I could fix it.

File: src/ConnectedMode/ServerSentEvents/SSESessionManager.cs [hunk 4]
```diff
@@ -67,25 +71,42 @@ public void Dispose()
...
+                if (bindingConfiguration.Equals(currentBindingConfiguration))
+                {
+                    return;
+                }
 
-            currentSession = sseSessionFactory.Create(bindingConfiguration.Project.ProjectKey);
+                EndCurrentSession();
 
-            currentSession.PumpAllAsync().Forget();
+                currentBindingConfiguration = bindingConfiguration;
```
Comment (rita-sonarsource) [inline src/ConnectedMode/ServerSentEvents/SSESessionManager.cs:89]: I think we should not cache this.
This solves a scenario that can only happen during first time initialization, and it is unlikely to happen in practice. If we don't cache, then worst case we will throw away the session and make a new one which is identical.
I think this is preferred as it simplifies the code and the impact is not large.

File: src/ConnectedMode.UnitTests/ServerSentEvents/SSESessionManagerTests.cs [hunk 5]
```diff
@@ -29,48 +30,85 @@ namespace SonarLint.VisualStudio.ConnectedMode.UnitTests.ServerSentEvents;
...
+
+    [TestMethod]
+    public void Ctor_WhenInConnectedModeOnCreation_CreatesSession()
+    {
+        var testScope = new TestScope(TestScope.CreateConnectedModeBindingConfiguration(DefaultProjectKey));
+        testScope.SetUpSSEFactoryToReturnNoOpSSESession(DefaultProjectKey);
+
+        var _ = testScope.CreateTestSubject();
+
+        testScope.SSESessionFactoryMock.Verify(factory => factory.Create(DefaultProjectKey), Times.Once);
```
Comment (rita-sonarsource) [inline src/ConnectedMode.UnitTests/ServerSentEvents/SSESessionManagerTests.cs:87]: I think we should also check that the session was started

File: src/ConnectedMode.UnitTests/ServerSentEvents/SSESessionManagerTests.cs [hunk 6]
```diff
@@ -29,48 +30,85 @@ namespace SonarLint.VisualStudio.ConnectedMode.UnitTests.ServerSentEvents;
...
+        testScope.SetUpSSEFactoryToReturnNoOpSSESession(DefaultProjectKey, () => isSubscribed.Should().BeTrue());
+
+        var _ = testScope.CreateTestSubject();
+
+        testScope.ActiveSolutionBoundTrackerMock.VerifyAdd(tracker => tracker.SolutionBindingChanged += It.IsAny<EventHandler<ActiveSolutionBindingEventArgs>>(), Times.Once);
+        testScope.SSESessionFactoryMock.Verify(factory => factory.Create(DefaultProjectKey), Times.Once);
+    }
+
+    [TestMethod]
+    public void Ctor_WhenInConnectedModeOnCreation_CreatesSession()
```
Comment (rita-sonarsource) [inline src/ConnectedMode.UnitTests/ServerSentEvents/SSESessionManagerTests.cs:80]: What about the standalone test? with explicit verification that session factory was not called.

File: src/ConnectedMode.UnitTests/ServerSentEvents/SSESessionManagerTests.cs [hunk 7]
```diff
@@ -97,16 +136,31 @@ public void OnSolutionChanged_WhenChangesFromConnectedToConnected_CancelsSession
         sessionMock2.Verify(session => session.Dispose(), Times.Never);
     }
 
+    [TestMethod]
+    public void OnSolutionChanged_WhenAttemptsToConnectToAlreadyConnectedProject_KeepsPreviousSessionAlive()
```
Comment (rita-sonarsource) [inline src/ConnectedMode.UnitTests/ServerSentEvents/SSESessionManagerTests.cs:140]: See my comment in the product code. I think this is dangerous.
The user can force a connection to restart by right-clicking in the team explorer and picking "Update". This will disable it.

File: src/ConnectedMode/ServerSentEvents/SSESessionManager.cs (hunk 4 above)
Comment (rita-sonarsource) [reply src/ConnectedMode/ServerSentEvents/SSESessionManager.cs:89]: Edit: I think it's also dangerous. The user can force an update by clicking on "Update" in the team explorer. I _think_ that doing that fires this event too. This will kill the force-update functionality.

**Review by rita-sonarsource — APPROVED** [review]
LGTM but I'd remove the caching of the bindingConfiguration

File: src/ConnectedMode/ServerSentEvents/SSESessionManager.cs (hunk 2 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode/ServerSentEvents/SSESessionManager.cs:39]: I hard disagree. What if the deadlock happens on a user's machine? How are you going to repro and fix it?

File: src/ConnectedMode/ServerSentEvents/SSESessionManager.cs (hunk 4 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode/ServerSentEvents/SSESessionManager.cs:89]: Will the active solution tracker even raise an event in this case? Since it's the same binding configuration, I'm pretty sure it's not going to

File: src/ConnectedMode.UnitTests/ServerSentEvents/SSESessionManagerTests.cs (hunk 7 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode.UnitTests/ServerSentEvents/SSESessionManagerTests.cs:140]: see my reply in that thread

File: src/ConnectedMode/ServerSentEvents/SSESessionManager.cs (hunk 4 above)
Comment (rita-sonarsource) [reply src/ConnectedMode/ServerSentEvents/SSESessionManager.cs:89]: You can check; put a breakpoint in the event and click on update.

File: src/ConnectedMode/ServerSentEvents/SSESessionManager.cs (hunk 2 above)
Comment (rita-sonarsource) [reply src/ConnectedMode/ServerSentEvents/SSESessionManager.cs:39]: It's a minor issue so agree to disagree and keep it however.
